pytorch_lightning/strategies/collaborative.py

```diff
@@ -0,0 +1,511 @@
+        ]
+
+        if self.endpoint:
+            self.host = self.host if self.host is not None else self.DEFAULT_HOST
+            self.port = self.port if self.port is not None else self.DEFAULT_PORT
+            self._start_server_process(self.host, self.port)
+            self._log_endpoint_helper_message(visible_addresses)
+        elif self.peer_endpoint:
+            log.info("Machine received initial peers from endpoint.")
+        elif self.initial_peers is None:
+            log.info(
+                f"\nOther machines can connect running the same command:\n"
+                f"INITIAL_PEERS={','.join(visible_addresses)} python ...\n"
+                "or passing the peers to the strategy:\n"
+                f"CollaborativeStrategy(initial_peers='{','.join(visible_addresses)}')"
+            )
+
+    def _log_endpoint_helper_message(self, visible_addresses: List[str]) -> None:
+        assert self.host is not None
+        resolved_host = self.host
+        if "0.0.0.0" in self.host:
+            # use the visible multi-addresses to figure out the IP that has been exposed
+            # todo (sean): this is pretty hacky, worth investigating.
+            p = re.compile(r"[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+")
+            # todo (sean): we select one address from here, could we have multiple?
+            resolved_host = {p.findall(maddr)[0] for maddr in visible_addresses}.pop()
+        log.info(
+            "\nSidecar endpoint enabled to serve peers.\n"
+            "Other peers can connect via:\n"
+            f"PEER_ENDPOINT={resolved_host}:{self.port} python ...\n"
+            "or pass the peer endpoint address to the strategy:\n"
+            f"CollaborativeStrategy(peer_endpoint='{resolved_host}:{self.port}')"
+        )
+
+    def _start_server_process(self, host: str, port: int) -> None:
+        dht = self.dht
+
+        class DHTHandler(BaseHTTPRequestHandler):
+            def do_GET(self) -> None:
+                """Respond to a GET request."""
+                self.send_response(200)
+                self.send_header("Content-type", "text/html")
+                self.end_headers()
+
+                visible_peers = [
+                    str(a) for a in dht.get_visible_maddrs() if not ipaddress.ip_address(a.values()[0]).is_loopback
+                ]
+
+                self.wfile.write("\n".join(visible_peers).encode())
+
+        server = http.server.ThreadingHTTPServer((host, int(port)), DHTHandler)
+        thread = threading.Thread(target=server.serve_forever)
+        thread.daemon = True
+        thread.start()
+
+    def _get_initial_peers_from_endpoint(self, retry_initial_peers: int, retry_peer_sleep_duration: int) -> List:
+        peers = None
+        for x in range(retry_initial_peers):
+            try:
+                peers = self._get_peers()
+                break
+            except Exception:
```

Comment:
do we know a specific type of exception that is being raised?
ConnectionError? ConnectionTimeout?
if it is known, we should catch the specific one(s), please